API/models/model.go

```diff
@@ -652,133 +614,175 @@ func DeleteEntity(entity string, id string, userRoles map[string]Role) *u.Error
+	var idFilter bson.M
+	if u.IsEntityNonHierarchical(u.EntityStrToInt(entityStr)) {
+		idFilter = bson.M{"slug": id}
+	} else {
+		idFilter = bson.M{"id": id}
+	}
+
+	result, err := WithTransaction(func(ctx mongo.SessionContext) (interface{}, error) {
+		//Update timestamp requires first obj retrieval
+		//there isn't any way for mongoDB to make a field
+		//immutable in a document
+		var oldObj map[string]any
+		var err *u.Error
+		if entityStr == u.HIERARCHYOBJS_ENT {
+			oldObj, err = GetHierarchyObjectById(id, u.RequestFilters{}, userRoles)
+			if err == nil {
+				entityStr = oldObj["category"].(string)
+			}
+		} else {
+			oldObj, err = GetObject(idFilter, entityStr, u.RequestFilters{}, userRoles)
+		}
+		if err != nil {
+			return nil, err
+		}
 
-	// Check user permissions in case domain is being updated
-	if entInt != u.DOMAIN && entInt < u.ROOMTMPL && (oldObj["domain"] != t["domain"]) {
-		if perm := CheckUserPermissions(userRoles, entInt, t["domain"].(string)); perm < WRITE {
+		entity := u.EntityStrToInt(entityStr)
+
+		// Check if permission is only readonly
+		if u.IsEntityHierarchical(entity) && oldObj["description"] == nil {
+			// Description is always present, unless GetEntity was called with readonly permission
 			return nil, &u.Error{Type: u.ErrUnauthorized,
 				Message: "User does not have permission to change this object"}
 		}
-	}
 
-	// Update database callback
-	callback := func(sessCtx mongo.SessionContext) (interface{}, error) {
-		mongoRes := GetDB().Collection(ent).FindOneAndReplace(ctx,
-			req, t,
-			&options.FindOneAndReplaceOptions{ReturnDocument: &retDoc})
+		// Update old object data with patch data
+		if isPatch {
+			var formattedOldObj map[string]interface{}
+			// Convert primitive.A and similar types
+			bytes, _ := json.Marshal(oldObj)
+			json.Unmarshal(bytes, &formattedOldObj)
+			// Update old with new
+			err := updateOldObjWithPatch(formattedOldObj, updateData)
+			if err != nil {
+				return nil, &u.Error{Type: u.ErrBadFormat, Message: err.Error()}
+			}
+
+			updateData = formattedOldObj
+			// Remove API set fields
+			delete(updateData, "id")
+			delete(updateData, "lastUpdated")
+			delete(updateData, "createdDate")
+		}
+
+		err = prepareUpdateObject(ctx, entity, id, updateData, oldObj, userRoles)
+		if err != nil {
+			return nil, err
+		}
+
+		mongoRes := repository.GetDB().Collection(entityStr).FindOneAndReplace(
+			ctx,
+			idFilter, updateData,
+			options.FindOneAndReplace().SetReturnDocument(options.After),
+		)
 		if mongoRes.Err() != nil {
 			return nil, mongoRes.Err()
 		}
-		if oldObj["id"] != t["id"] {
+
+		if oldObj["id"] != updateData["id"] {
 			// Changes to id should be propagated to its children
-			if err := PropagateParentIdChange(ctx, oldObj["id"].(string),
-				t["id"].(string), entInt); err != nil {
+			err := PropagateParentIdChange(
```

Comment:
A change to the name of the domain should also propagate to all objects that have it or the name should not be modifiable.

Reply:
To be honest I don't understand this comment, here I only added enters between the parameters because the line was too long, but I didn't change anything in the logic.

Reply:
It was something I noticed testing this branch but it was already there, I put the comment to not forget it, I'll create a separate issue for it